File: algorithms/cracking_the_coding_interview/tries_contacts.php
<?php
declare(strict_types=1);

const OPERATION_ADD = 'add';
const OPERATION_FIND = 'find';

class Trie
{
    /** @var int */
    private $numberOfWords = 0;

    /** @var Trie[] [$char => $trie...] */
    private $children = [];

    /**
     * @param string $word
     */
    public function add(string $word)
    {
        $this->numberOfWords++;

        if (empty($word)) {
            return;
        }

        $firstChar = $word[0];
        $word = substr($word, 1);
        if (empty($this->children) || empty($this->children[$firstChar])) {
            $child = new Trie();
            $this->children[$firstChar] = $child;
            $child->add($word);
        } else {
            $this->children[$firstChar]->add($word);
        }

    }

    /**
     * @param string $word
     * @return int
     */
    public function find(string $word): int
    {
        if (empty($word)) {
            return $this->numberOfWords;
        }

        $firstChar = $word[0];
        if (empty($this->children[$firstChar])) {
            return 0;
        }

        return $this->children[$firstChar]->find(substr($word, 1));
    }
}

$handle = fopen ('php://stdin', 'r');
fscanf($handle,'%d',$n);

$trie = new Trie();
$dictionary = [];

for ($i = 0; $i < $n; $i++) {
    fscanf($handle, '%s %s', $op, $contact);

    if ($op === OPERATION_ADD && empty($dictionary[$contact])) {
        $trie->add($contact);
        $dictionary[$contact] = true;
    } elseif ($op === OPERATION_FIND) {
        echo $trie->find($contact) . "\n";
    }
}

File: algorithms/cracking_the_coding_interview/tries_contacts.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const (
	OperationAdd  = "add"
	OperationFind = "find"
)

type Trie struct {
	numberOfWords int
	children      map[rune]*Trie
}

func NewTrie() *Trie {
	return &Trie{
		children: map[rune]*Trie{},
	}
}

func (t *Trie) add(word string) {
	t.numberOfWords++

	if len(word) == 0 {
		return
	}

	firstChar := []rune(word)[0]
	word = word[1:]
	child, ok := t.children[firstChar]
	if !ok {
		newChild := NewTrie()
		t.children[firstChar] = newChild
		newChild.add(word)
	} else {
		child.add(word)
	}
}

func (t Trie) find(word string) int {
	if len(word) == 0 {
		return t.numberOfWords
	}

	firstChar := []rune(word)[0]
	child, ok := t.children[firstChar]
	if !ok {
		return 0
	}

	return child.find(word[1:])
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	reader.ReadString('\n')

	trie := NewTrie()
	dictionary := make(map[string]bool)

	for {
		line, err := reader.ReadString('\n')
		line = strings.TrimSpace(line)
		lineData := strings.Fields(line)

		operation := lineData[0]
		contact := lineData[1]

		if _, ok := dictionary[contact]; !ok && operation == OperationAdd {
			trie.add(contact)
			dictionary[contact] = true
		} else if operation == OperationFind {
			fmt.Println(trie.find(contact))
		}

		if err != nil {
			break
		}
	}
}
